z_test.go

```diff
@@ -3185,3 +3193,227 @@ func TestOpenCloseLob(t *testing.T) {
 		})
 	}
 }
+
+func TestPreFetchQuery(t *testing.T) {
+
+	if os.Getenv("GODROR_TEST_SYSTEM_USERNAME") == "" ||
+		(os.Getenv("GODROR_TEST_SYSTEM_PASSWORD") == "") {
+		t.Skip("Please define GODROR_TEST_SYSTEM_USERNAME and GODROR_TEST_SYSTEM_PASSWORD env variables")
+	}
+	var err error
+	if testSystemDb, err = sql.Open("godror", testSystemConStr); err != nil {
+		panic(errors.Errorf("%s: %+v", testConStr, err))
+	}
+
+	ctx, cancel := context.WithTimeout(testContext("TestPreFetchQuery"), 30*time.Second)
+	defer cancel()
+
+	// Create a table used for Prefetch, ArrayFetch queries
+
+	tbl := "t_employees" + tblSuffix
+	testDb.ExecContext(ctx, "DROP TABLE "+tbl)
+	if _, err := testDb.ExecContext(ctx, "CREATE TABLE "+tbl+" (employee_id NUMBER, last_name VARCHAR2(30))"); err != nil {
+		t.Fatal(err)
+	}
+	defer testDb.Exec("DROP TABLE " + tbl)
+
+	const num = 120 // 120 rows to be created
+	nums := make([]godror.Number, num)
+	strs := make([]string, num)
+	for i := range nums {
+		nums[i] = godror.Number(strconv.Itoa(i))
+		strs[i] = fmt.Sprintf("%x", i)
+	}
+
+	tx, err := testDb.BeginTx(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tx.Rollback()
+	for i, tc := range []struct {
+		Name  string
+		Value interface{}
+	}{
+		{"employee_id", nums},
+		{"last_name", strs},
+	} {
+		res, execErr := tx.ExecContext(ctx,
+			"INSERT INTO "+tbl+" ("+tc.Name+") VALUES (:1)", //nolint:gas
+			tc.Value)
+		if execErr != nil {
+			t.Fatal("%d. INSERT INTO "+tbl+" (%q) VALUES (%+v): %#v", //nolint:gas
+				i, tc.Name, tc.Value, execErr)
+		}
+		ra, raErr := res.RowsAffected()
+		if raErr != nil {
+			t.Error(raErr)
+		} else if ra != num {
+			t.Errorf("%d. %q: wanted %d rows, got %d", i, tc.Name, num, ra)
+		}
+	}
+	tx.Commit()
+	sid := func() uint {
+		var sid uint
+		sql := "SELECT sys_context('userenv','sid') FROM dual"
+		err := testDb.QueryRow(sql).Scan(&sid)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Get godror version number
+		var driver string
+		sql2 := `SELECT UNIQUE client_driver
+         FROM v$session_connect_info
+         WHERE sid = :sid`
+		err = testDb.QueryRow(sql2, sid).Scan(&driver)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Log(driver)
+
+		return sid
+
+	}
+
+	// verify round trips for SingleRowFetch and MultiRowFetch
+	// and return failure on unexpected roundtrips
+
+	srt, mrt := runPreFetchTests(t, sid(), -1, -1) // prefetch, arraysize not given
```

Comment:
Thinking more about this, there will need to be a way to set the parameters to negative numbers (instead of using -1 to indicate 'use the default'). This is needed so error handling of invalid values can be tested.

Reply:
Maybe make something like -99 mean use the default.  This lets -1 be used in boundary tests to make sure error handling is correct.